### PR TITLE
`DataFrame`: Simple translations

### PR DESCRIPTION
**Pull request by vincentarelbundock**
Translated and tested:

* `DataFrame`: `max`, `mean`, `median`, `min`, `sum`, `var`, `std`, `first`, `last`, `head`, `tail`, `reverse`, `slice`, `null_count`, `estimated_size`
* `LazyFrame`: `max`, `mean`, `median`, `min`, `sum`, `var`, `std`, `first`, `last`, `head`, `tail`, `reverse`, `slice`
* `GroupBy`: `max`, `mean`, `median`, `min`, `sum`, `var`, `std`, `first`, `last`, `null_count`

TODO:

+ [ ] `.quantile`: `QuantileInterpolOptions` input
+ [ ] `is_duplicated()`: How do we return a `ChunkedArray` to R output?

### CODE REVIEW

**Comment by sorhawell** [comment]
Interesting that DataFrame.first and DataFrame.last is not implemented for py-polars, but they are in rust-polars just as you wrap them.

**Comment by sorhawell** [comment]
>How should I re-write the slice function below to make length argument optional (None)?

you can do like this
image

when I get lost in type conversions, I find it helpful to break up the lines and let rust-analyzer explain what types I have currently made

image

**Comment by sorhawell** [comment]
> Missing robj_to_u8 makes rob_to fail when I try to implement DataFrame.std()

you can cut a corner and use the extendr conversion like I [did here](https://github.com/pola-rs/polars/blob/f566963f526a11585805088c96e579045a0a2b79/py-polars/src/dataframe.rs#L1187) in the beginning.

we can make this PR into a co-op and then I can add the u8 conversion OR just use extendr in this PR, and I file an issue to upgrade u8 conversions.

**Comment by sorhawell** [comment]
[here's a minimal example + tidying up the GroupBy class](https://github.com/pola-rs/r-polars/commit/8f73892dca58fb4f479de264d84435282df56daf)

after zzz.R is updated you can do like this

```R
GroupBy_max = function() {
  self$agg(pl$all()$max())
}
```

**Comment by vincentarelbundock** [comment]
Thanks for all the extra info! Everything seems to have worked without a hitch.

I used the `u8` shortcut for now.

I'll look at a few more candidates for translation and let you know when I think this PR is ready for a review.

**Comment by vincentarelbundock** [comment]
@sorhawell This PR now covers most of the trivial translations for `DataFrame`, `LazyFrame`, and `GroupBy`. I think it's a good place to stop for now.

Let me know when you've had time to look at this and I can implement any change you need.

Once this is merged, I'll turn to the S3 methods PR https://github.com/pola-rs/r-polars/pull/107.

Depending on time, energy, etc., I might eventually come back to translate some of the trickier bits, but those will have to wait a bit.

**Review by eitsupi — COMMENTED** [review]
Thanks for working on this!
If you don't mind, would you consider improving the test using patrick?

File: tests/testthat/test-dataframe.R
[the remark this answers is not in the document]
Comment (vincentarelbundock) [reply tests/testthat/test-dataframe.R:1]: OK: 3582ca1803d86d45d8412ebca20dd955049081fe

**Comment by sorhawell** [comment]
> Sorry for the dirty commit history. Made some stupid mistakes with `patrick`, but it should work now.

no problemo, any PR gets squashed anyways :)

**Comment by sorhawell** [comment]
looks very good :) I will add to this PR some minor style changes later, I was just derping around in how to commit to a PR

File: R/dataframe__frame.R [hunk 1]
```diff
@@ -864,3 +864,133 @@ DataFrame_unnest = function(names = NULL) {
...
+#' @param length integer or NULL
+#' @examples 
+#' pl$DataFrame(mtcars)$slice(2, 4)
+#' mtcars[2:6,]
+DataFrame_slice = function(offset, length = NULL) {
+  self$lazy()$slice(offset, length)$collect()
+}
+
+
+#' @title Null count
```
Comment (sorhawell) [inline R/dataframe__frame.R:975]: missing a few tests for null_count and estimated_size

**Comment by sorhawell** [comment]
I added some changes to the non-fallible rust methods. These do not need to return a Result, and if default arg value on R side, the extendr wrapper can be used directly as is with flag "use_extendr_wrapper".


with a few tests for null_count and estimated_size, this looks like a wrap :)

File: src/rust/src/rdataframe/mod.rs
[the remark this answers is not in the document]
Comment (sorhawell) [reply src/rust/src/rdataframe/mod.rs:1]: ... no change needed

**Comment by sorhawell** [comment]
patrick tests look beautiful

**Comment by vincentarelbundock** [comment]
Great! I will study your changes and comments, and will make changes as appropriate. It may take me a few days because of work + family stuff.

**Comment by vincentarelbundock** [comment]
I pushed a simple test for `.estimated_size()`.

Here’s something weird: `.as_data_frame()` changes the values in the `.null_count()` data frame.

Install the PR branch:

``` r
remotes::install_github("vincentarelbundock/r-polars@translation")
```

``` r
library(rpolars)
tmp = mtcars
tmp[1:2, 1:2] = NA
tmp[5, 3] = NA
a = pl$DataFrame(tmp)$null_count()

# correct
a
# polars DataFrame: shape: (1, 11)
# ┌─────┬─────┬──────┬─────┬─────┬─────┬─────┬──────┬──────┐
# │ mpg ┆ cyl ┆ disp ┆ hp  ┆ ... ┆ vs  ┆ am  ┆ gear ┆ carb │
# │ --- ┆ --- ┆ ---  ┆ --- ┆     ┆ --- ┆ --- ┆ ---  ┆ ---  │
# │ u32 ┆ u32 ┆ u32  ┆ u32 ┆     ┆ u32 ┆ u32 ┆ u32  ┆ u32  │
# ╞═════╪═════╪══════╪═════╪═════╪═════╪═════╪══════╪══════╡
# │ 2   ┆ 2   ┆ 1    ┆ 0   ┆ ... ┆ 0   ┆ 0   ┆ 0    ┆ 0    │
# └─────┴─────┴──────┴─────┴─────┴─────┴─────┴──────┴──────┘

# incorrect
a$as_data_frame()
#             mpg           cyl          disp hp drat wt qsec vs am gear carb
# 1 9.881313e-324 9.881313e-324 4.940656e-324  0    0  0    0  0  0    0    0
```

Any ideas why we get this?

**Comment by eitsupi** [comment]
Sorry, I merged #112 because NEWS was not updated correctly.
Could you please move the news item to the correct position?

**Comment by vincentarelbundock** [comment]
I just reverted the NEWS commit. Not sure what's the best process to add it back, but for the record, here is what I had written:

```markdown
- New methods implemented for DataFrame, LazyFrame, and GroupBy objects: min, max, mean, median, sum, std, var, first, last, head, tail, reverse, slice, null_count, estimated_size (#105 @vincentarelbundock)

## New Contributors

- @grantmcdermott made their first contribution in #81
- @vincentarelbundock made their first contribution in #105
```

**Comment by vincentarelbundock** [comment]
Ah I see what you mean @eitsupi . Thanks for the merge. I updated NEWS in the right position.

**Comment by vincentarelbundock** [comment]
nope, still broken. Giving up on this now. We can update the NEWS at the very end.

**Comment by sorhawell** [comment]
> Here’s something weird: .as_data_frame() changes the values in the .null_count() data frame.

Currently  r-polars maps  arrow-u32 to R bit64::nteger64 which again is a hack to get i64 in R. Under the hood bit64::integer64 is just an i64 which is tagged with "SexpReal", if bit64 is not loaded the user is in for a surprise.

rpostgresql on other packages does the same. I prefer in future to map to f64 and let the connoisseurs actively  opt for bit64 package

```R

> library(bit64)
> library(rpolars)
> tmp = mtcars
> tmp[1:2, 1:2] = NA
> tmp[5, 3] = NA
> a = pl$DataFrame(tmp)$null_count()
> a
polars DataFrame: shape: (1, 11)
┌─────┬─────┬──────┬─────┬─────┬─────┬─────┬──────┬──────┐
│ mpg ┆ cyl ┆ disp ┆ hp  ┆ ... ┆ vs  ┆ am  ┆ gear ┆ carb │
│ --- ┆ --- ┆ ---  ┆ --- ┆     ┆ --- ┆ --- ┆ ---  ┆ ---  │
│ u32 ┆ u32 ┆ u32  ┆ u32 ┆     ┆ u32 ┆ u32 ┆ u32  ┆ u32  │
╞═════╪═════╪══════╪═════╪═════╪═════╪═════╪══════╪══════╡
│ 2   ┆ 2   ┆ 1    ┆ 0   ┆ ... ┆ 0   ┆ 0   ┆ 0    ┆ 0    │
└─────┴─────┴──────┴─────┴─────┴─────┴─────┴──────┴──────┘
> as.data.frame(a)
  mpg cyl disp hp drat wt qsec vs am gear carb
1   2   2    1  0    0  0    0  0  0    0    0
> str(as.data.frame(a))
'data.frame':	1 obs. of  11 variables:
 $ mpg :integer64 2 
 $ cyl :integer64 2 
 $ disp:integer64 1 
 $ hp  :integer64 0 
 $ drat:integer64 0 
 $ wt  :integer64 0 
 $ qsec:integer64 0 
 $ vs  :integer64 0 
 $ am  :integer64 0 
 $ gear:integer64 0 
 $ carb:integer64 0 

```

**Comment by vincentarelbundock** [comment]
Got it, thanks. I added very simple .null_count() and .estimated_size() tests.

AFAICT, this completes my TODO list. Feel free to let me know; I'm happy to make more changes to this PR if helpful.

**Comment by sorhawell** [comment]
excellent many thanks @vincentarelbundock

**Comment by vincentarelbundock** [comment]
Thanks for your guidance, edits, and prompt responses. This was a great contribution experience :)

**Pull request merged**